visualization/app/codeCharta/ui/customViews/customViewFileStateConnector.ts

```diff
@@ -0,0 +1,69 @@
+"use strict"
+import {FileSelectionState, FileState} from "../../model/files/files";
+import {CustomViewMapSelectionMode} from "../../model/customView/customView.api.model";
+import md5 from "md5";
+
+export class CustomViewFileStateConnector {
+
+    private readonly files: FileState[] = []
+    private fileNameParts: string[] = []
+    private mapChecksums: string[] = []
+    private mapSelectionMode: CustomViewMapSelectionMode = CustomViewMapSelectionMode.SINGLE
+
+    constructor(files: FileState[]) {
+        this.files = files
+
+        if (typeof this.files !== 'undefined') {
+            this.processFiles()
+        }
+    }
+
+    private processFiles() {
+        for (const file of this.files) {
+            if (file.selectedAs !== FileSelectionState.None) {
+                this.setMapSelectionMode(file.selectedAs)
+                this.mapChecksums.push(md5(JSON.stringify(file.file.map)))
```

Comment:
How expensive is it to stringify and then to calculate the md5 hash on the result?
I'm just thinking maybe we can call this operation somewhere else, where we already work with the JSON, to reduce the potential overhead? 
Please correct me if I'm wrong here

Reply:
The has value should always be calculated for a file, not only for a map. Ideally the file contains a precomputed checksum. If it does not (e.g., old cc.json files) calculate it on the fly.

Reply:
I have had a look at this. We can calculate missing checksums on the fly, if a file has been choosen. We should also adjust the FileDownload, so that we generate a fresh checksum for the downloaded .cc.json file (not to forget: the parsers which generate .cc.json files as well).

For which properties the checksum would be built? Propably for all properties of the `ExportCCFile` Type (including Blacklist, MarkedPackges, etc.).
This will have negative consequences for existing CustomViews. As soon if the blacklist will change and the .cc.json file will be downloaded, all corresponding CustomViews are outdated for the loaded map and for the new downloaded file as well. Note that a CustomView can override the blacklist anyway. This is not very user friendly.

Solution? We could build a checksum during the download process based on the map nodes, projectName, apiVersion, attributeTypes and edges. If a file is uploaded, missing checksums could be generated on the fly.
But that checksum would not be a generic checksum! It would be a kind of "CustomViewChecksum" or anything like that.

We could add a new optional generic `attributes` object as a property of the `ExportCCFile` type. There we could store some kind of meta attributes for the .cc.json file. For example `attributes['customViewChecksum']` could hold the md5 hash of the map nodes, edges, apiVersion, and projectName (but not the blacklist and markedPackages). Any suggestions appreciated.

Reply:
Please ignore the current download possibility. In a later stage, the download button should be completely replaced and the cc.json file is not changed at all (and should not be downloaded). Instead, only a configuration file would be downloaded that corresponds with the original cc.json file.